compliance-monitor/monitor.py

```diff
@@ -176,38 +190,50 @@ def import_bootstrap(bootstrap_path, conn):
         conn.commit()
 
 
-def import_cert_yaml(yaml_path, conn):
+def precompute_targets(version):
+    suite = compile_suite(f"{version['version']}", version['include'])
+    return {
+        'suite': suite,
+        'targets': {
+            tname: suite.select(tname, target_spec)
+            for tname, target_spec in version['targets'].items()
+        }
+    }
+
+
+def import_cert_yaml(yaml_path, target_dict):
     yaml = ruamel.yaml.YAML(typ='safe')
     with open(yaml_path, "r") as fileobj:
-        document = yaml.load(fileobj.read())
-    # The following will also delete entries that are not present in the given yaml.
-    # It is paramount that all extant primary keys be kept because the reports refer to them, and
-    # deletions will cascade! But we should only ever delete entries for non-stable versions; stable versions
-    # are deemed immutable except maybe for checks, and for those, at least the ids are immutable.
-    with conn.cursor() as cur:
-        scopeid = db_update_scope(cur, document)
-        all_versions = set()
-        for vdata in document['versions']:
-            versionid = db_update_version(cur, scopeid, vdata)
-            all_versions.add(versionid)
-            all_standards = set()
-            for sdata in vdata['standards']:
-                standardid = db_update_standard(cur, versionid, sdata)
-                all_standards.add(standardid)
-                all_checks = set()
-                for cdata in sdata.get('checks', ()):
-                    checkid = db_update_check(cur, versionid, standardid, cdata)
-                    all_checks.add(checkid)
-                db_filter_checks(cur, standardid, lambda checkid, *_: checkid in all_checks)
-            db_filter_standards(cur, versionid, lambda standardid, *_: standardid in all_standards)
-        db_filter_versions(cur, scopeid, lambda versionid, *_: versionid in all_versions)
-        conn.commit()
+        spec = load_spec(yaml.load(fileobj.read()))
+    annotate_validity(spec['timeline'], spec['versions'], date.today())
+    target_dict[spec['uuid']] = {
+        'spec': spec,
+        'versions': {
+            version['version']: precompute_targets(version)
+            for version in spec['versions'].values()
+        }
+    }
 
 
-def import_cert_yaml_dir(yaml_path, conn):
+def import_cert_yaml_dir(yaml_path, target_dict):
     for fn in sorted(os.listdir(yaml_path)):
         if fn.startswith('scs-') and fn.endswith('.yaml'):
-            import_cert_yaml(os.path.join(yaml_path, fn), conn)
+            import_cert_yaml(os.path.join(yaml_path, fn), target_dict)
+
+
+def get_scopes():
+    """returns thread-local copy of the scopes dict"""
```

Comment:
Again, is there anything that speaks against using the high-level API? https://docs.python.org/3/library/threading.html#thread-local-data

Reply:
just FYI, I had a look at how `threading.local` is implemented, and it's beyond bloated (as if Python in itself wasn't already beyond bloated)